hedgehog/src/Hedgehog/Internal/Runner.hs

```diff
@@ -126,11 +127,20 @@ checkReport ::
   -> Test m ()
   -> (Report -> m ())
   -> m Report
-checkReport cfg size0 seed0 test0 updateUI =
-  let
-    test =
-      catchAll test0 (fail . show)
+checkReport cfg size0 seed0 test updateUI =
+  checkReportNoCatch cfg size0 seed0 (catchAll test (fail . show)) updateUI
 
+checkReportNoCatch ::
+     forall m.
+     MonadIO m
+  => PropertyConfig
+  -> Size
+  -> Seed
+  -> Test m ()
+  -> (Report -> m ())
+  -> m Report
+checkReportNoCatch cfg size0 seed0 test updateUI =
```

Comment:
A simple Hspec with QuickCheck program to test this:
```
{-# LANGUAGE ScopedTypeVariables #-}
import           Data.Word
import           Test.Hspec
import           Test.Hspec.QuickCheck

main :: IO ()
main = hspec $
  prop "exception handling" $ \ (i :: Word8) -> 
    throwError i `shouldThrow` errorCall ("error " ++ show i)

throwError :: Word8 -> IO a
throwError i = do
  let msg = "error " ++ show i
  putStrLn msg
  error msg
```
shows that `throwError` does indeed get called more than once.

Its actually going to be the same for Hedgehog because its the `Test.Hspec.QuickCheck.prop` function that makes sure that the condition is evaluated more than once. I just need to implement my own `Test.Hspec.Hedgehog.prop` function.

So it seems that what Hedgehog needs to expose is some way of allowing a property to be tested just once, but with a test state that is maintained over multiple calls.